docs/source/deep_dives/sample_packing.rst

```diff
@@ -0,0 +1,247 @@
+
+      * Sample packing concatenates multiple data samples into a single packed
+      sample up to the maximum sequence length to boost model training throughput.
+      * Enabling sample packing for the Stanford Alpaca dataset yielded up to a
+      5x boost in throughput (~350 tokens/sec/gpu -> ~1800 tokens/sec/gpu)
+      * Adding in-sample position encoding and masking to eliminate cross-attending
+      of unrelated samples improved fine-tuned models on hellaswag (+2%) compared
+      to packing without in-sample masking, but showed no statistically significant
+      difference on truthfulqa_mc2.
+
+
+Background
+----------
+Context windows for modern LLMs are ever-increasing, from Llama3 at 8k to
+Gemini 1.5 Pro which is pushing 2 million. In practice, many OSS text datasets
+(such as Stanford Alpaca) consist of mostly short sequence length samples,
+averaging at ~100-1000, which is a small fraction of the full context. This can
+lead to significant padding compute and memory overhead, even if ignored or masked
+out in the loss function. Sample packing minimizes this overhead by jamming as
+many samples as possible in the extra space in the context window. This can massively
+increase throughput for model training, since the model will see more data per step.
+However, there are several caveats to consider when enabling sample packing for a
+transformer decoder model.
+1. When concatenating multiple samples in a single context window, we need a way
+to tell the model to not attend to all other previous samples and only attend to
+the sample the current token belongs to. This can be achieved with a block causal
+mask to prevent cross-contamination.
+2. Similarly, for position encodings we need to encode relative position of a
+token within its sample instead of absolute position across all samples in a
+single context window.
+3. Ideally, the block causal mask and relative position encoding is compatible
+with flash attention to maximize speed-ups.
+
+Masking to avoid cross-contamination during fine-tuning
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+By default, a causal mask is passed into attention calculations for a transformer
+decoder so that each token can attend to all previous tokens but not any future
+tokens. This is sound for a single sample, where each token requires the context
+provided by all the previous. However, in a packed sample which may contain numerous
+different samples only separated by EOS tokens, all consecutive samples should not
+attend to previous samples. Instead, we can create a block causal mask where each
+sample within a pack has its own causal mask. This way the packs are less noisy
+to the model as it only has to focus on attending each sample to itself.
+
+[ADD MASK PICTURE]
+
+Block causal masks have been implemented in several publications (ref and ref) as
+a way to directly address issues with cross-contamination. Some users have observed
```

Comment:
from section C: Related work in the above paper, they mention

> For instance, GPT-3  is trained with a maximum sequence length of 2048 where a single sequence may contain multiple segments of sentences separated by a special end of segment token ... GPT-3 does
not mask the attention between different segments in a sequence

and

>  RoBERTa and many other models in production like T5 [24 ] use a similar packing approach as GPT-3